src/classes/TEST_RecurringDonationBuilder.cls

```diff
@@ -579,6 +602,32 @@ public without sharing class TEST_RecurringDonationBuilder {
         return rd;
     }
 
+    /***
+    * @description Retrieves the value of CreatedDate from valuesByFieldName and removes the key
+    * @return String
+    */
+    private String getCreatedDateAndRemoveKey() {
+        String createdOn;
+        if (valuesByFieldName.containsKey('CreatedDate')) {
+            createdOn = ((Datetime) valuesByFieldName.get('CreatedDate')).format('yyyy-MM-dd\'T\'HH:mm:ss.SSS+0000');
+            valuesByFieldName.remove('CreatedDate');
+        }
+        return createdOn;
+    }
+
+    /***
+    * @description Inserts Created Date into Recurring Donation record
+    * @param rd npe03__Recurring_Donation__c
+    * @param createdOn String representing Datetime
+    * @return npe03__Recurring_Donation__c
+    */
+    private npe03__Recurring_Donation__c handleCreatedDate(npe03__Recurring_Donation__c rd, String createdOn) {
```

Comment:
The `handleCreatedDate()` can call `getCreatedDate()` and if `isBlank(createdDate)` then just return `rd`, otherwise the rest of logic here applies. Thus no specific logic about created date is required to be checked in the `build()`.

Reply:
I don't believe that will work.  CreatedDate needs to be removed before the rest of the query is built, so if we are going to remove the CreatedDate value and mock it with serialize/deserialize, I think we have to know it is there and remove it before the preceding block executes.  I think the two parts of this need to be split.

Reply:
I agree we need to check if the CreatedDate should be set. The question is who/where. If the current `handleCreatedDate()` is called from another method, that another method would need to check if the created date should be set, so repetition of the code. Thus, it would be best if the method itself validates and determines if the created date should be set.

```
public npe03__Recurring_Donation__c build() {
        npe03__Recurring_Donation__c rd = new npe03__Recurring_Donation__c();

        for (String fieldName : valuesByFieldName.keySet()) {
            rd.put(fieldName, valuesByFieldName.get(fieldName));
        }

        rd = handleCreatedDate(rd);

        if (calculateNextDonationDate) {
            new RD2_RecurringDonation(rd)
                .reviseNextDonationDateBeforeInsert(new RD2_ScheduleService());
        }

        return rd;
    }

    /***
    * @description Sets CreatedDate (if provided) on the Recurring Donation record
    * @param rd npe03__Recurring_Donation__c
    * @return npe03__Recurring_Donation__c
    */
    private npe03__Recurring_Donation__c handleCreatedDate(npe03__Recurring_Donation__c rd) {
        final String fieldName = 'CreatedDate';

        if (!valuesByFieldName.containsKey(fieldName)) {
            return rd;
        }

        Datetime createdDate = ((Datetime) valuesByFieldName.get(fieldName));
        valuesByFieldName.remove(fieldName);

        String rdRecord = JSON.serialize(rd);
        rdRecord = rdRecord.left(rdRecord.length() - 1);
        rdRecord += ',"' + fieldName + '":"' + createdDate.format('yyyy-MM-dd\'T\'HH:mm:ss.SSS+0000') + '"}';
        return (npe03__Recurring_Donation__c) JSON.deserialize(rdRecord, npe03__Recurring_Donation__c.class);
    }
```

Reply:
I see what you mean, the CreatedDate needs to be removed, then RD built, then Created date set. I was changing the method as part of the code review and getting errors. Here is one version of the code that just removes the field and then passes it to the handler to check if and how to set the field. See if this will work and if not, feel free to leave the code as is. Tis version does not need `getCreatedDateAndRemoveKey()`

```
public npe03__Recurring_Donation__c build() {
        npe03__Recurring_Donation__c rd = new npe03__Recurring_Donation__c();

        //Get the Created Date and remove it from the field list
        //since it is not an editable field
        Object createdDate = valuesByFieldName.remove('CreatedDate');

        //Set editable fields on the Recurring Donation 
        for (String fieldName : valuesByFieldName.keySet()) {
            rd.put(fieldName, valuesByFieldName.get(fieldName));
        }

        //Set Created Date on the Recurring Donation if the value is provided
        rd = handleCreatedDate(rd, createdDate);

        if (calculateNextDonationDate) {
            new RD2_RecurringDonation(rd)
                .reviseNextDonationDateBeforeInsert(new RD2_ScheduleService());
        }

        return rd;
    }

    /***
    * @description Sets CreatedDate (if provided) on the Recurring Donation record
    * @param rd Recurring Donation record
    * @param createdDate Created date
    * @return npe03__Recurring_Donation__c
    */
    private npe03__Recurring_Donation__c handleCreatedDate(npe03__Recurring_Donation__c rd, Object createdDate) {
        if (createdDate == null) {
            return rd;
        }

        String formattedDate = ((Datetime) createdDate).format('yyyy-MM-dd\'T\'HH:mm:ss.SSS+0000');

        String rdRecord = JSON.serialize(rd);
        rdRecord = rdRecord.left(rdRecord.length() - 1);
        rdRecord += ',"' + 'CreatedDate' + '":"' + formattedDate + '"}';
        return (npe03__Recurring_Donation__c) JSON.deserialize(rdRecord, npe03__Recurring_Donation__c.class);
    }
```

Reply:
@RadSFDeveloper @andrewyu-salesforce  I refactored the code to use a class variable instead of the field/value map.  I think it makes the code much cleaner and addresses the issues you have raised.  Please take a look and see what you think.